nix-prefetch-scripts: add nix-prefetch-fossil

siraben opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

This is so that we can have update.passthru scripts for packages like `althttpd`.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

philiptaron commented:
I added @maggesi who is the maintainer for Fossil. If they sign off I'm happy to merge.

philiptaron commented:
Apologies on my inattention to this. I look forward to what you produce.

siraben committed:
nix-prefetch-scripts: add nix-prefetch-fossil

Add a nix-prefetch-fossil script that follows the same pattern as the
existing nix-prefetch-git, nix-prefetch-hg, etc. scripts. This allows
users to easily fetch Fossil repositories and calculate their hashes
for use with fetchfossil in Nix expressions.

The script supports:
- Fetching from Fossil repository URLs
- Specifying revisions (defaults to "trunk")
- Checking expected hashes
- JSON output compatible with other nix-prefetch scripts

siraben committed:
althttpd: add updateScript

siraben committed:
althttpd: unstable-2023-08-12 -> unstable-2025-08-22

siraben committed:
pikchr: add update script

siraben committed:
pikchr: unstable-2025-02-28 -> unstable-2025-05-12

philiptaron commented:
Needs to pass [`treefmt`](https://github.com/NixOS/nixpkgs/actions/runs/16610803214/job/47009178944?pr=407091).

siraben commented:
@philiptaron done

siraben commented:
@toastal addressed your comments, let me know what you think.

toastal reviewed (commented):
@siraben Q: what did your research return on using `[ -t 2 ]; … >&2` for non-errors? Will this fail downstream with `set -eu pipefail` + `nix-prefetch-fossil … | jq …`?

siraben commented:
Actually, let's get rid of `[ -t 2 ]`, in a pipe, stderr gets redirected, so it's false, the diagnostic messages won't be seen at all.

siraben commented:
@toastal no it won't fail with `pipefail` since stderr output doesn't affect pipeline exit codes.

toastal reviewed (commented):
Other than what I think are missed `""`, it looks good to me.

siraben commented on pkgs/build-support/fetchfossil/nix-prefetch-fossil:
```diff
@@ -0,0 +1,125 @@
+rm -f "$checkoutDir/.fslckout"
+
+# Calculate the hash
+hash=$(nix-hash --type "$hashType" --base32 "$checkoutDir")
```
Quotes are only needed if you use `$(...)` as an argument to a command
